File: 75 day challenge/day 05/code.py


class Solution:
    def reverseVowels(self, s: str) -> str:
        ans = list(s)
        vowels = "aeiouAEIOU"
        start, end = 0, len(ans)-1

        while start < end:
            if ans[start] not in vowels:
               start += 1
            if ans[end] not in vowels:
               end -= 1
            if ans[start] in vowels and ans[end] in vowels:
                ans[start], ans[end] = ans[end], ans[start]
                start += 1
                end -= 1
        return ''.join(ans)
               
        


def main():
    test_cases = [
        ("hello", "holle"),
        ("leetcode", "leotcede"),
        ("HELLO", "HOLLE"),
        ("HeLlo", "HoLle"),
        ("bcdfgh", "bcdfgh"),
        ("", "")
    ]
    solution = Solution()

    for s, expected_output in test_cases:
        result = solution.reverseVowels(s)
        print(f"Input: {s}, Output: {result}, Expected: {expected_output}")
        assert result == expected_output, f"Test failed for input: {s}"

if __name__ == "__main__":
    main()

File: 75 day challenge/day 05/code.cpp
#include <iostream>
#include <cassert>
#include <unordered_set>

using namespace std;

class Solution
{
public:
    bool isVowel(char x)
    {
        static const std::unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
        return vowels.find(x) != vowels.end();
    }
    string reverseVowels(string s)
    {
        int start = 0, end = s.size();
        while (start < end)
        {
            if (!(isVowel(s[start])))
            {
                start++;
            }
            if (!(isVowel(s[end])))
            {
                end--;
            }
            if (isVowel(s[start]) and isVowel(s[end]))
            {
                char temp = s[start];
                s[start] = s[end];
                s[end] = temp;
                start++;
                end--;
            }
        }
        return s;
    }
};

void runTests()
{
    struct TestCase
    {
        string input;
        string expected_output;
    };

    TestCase test_cases[] = {
        {"hello", "holle"},
        {"leetcode", "leotcede"},
        {"HELLO", "HOLLE"},
        {"HeLlo", "HoLle"},
        {"bcdfgh", "bcdfgh"},
        {"", ""}};

    for (const auto &test_case : test_cases)
    {
        Solution solution;
        string result = solution.reverseVowels(test_case.input);
        cout << test_case.input << " " << result << " " << test_case.expected_output << endl;
        assert(result == test_case.expected_output);
    }
}

int main()
{
    runTests();
    return 0;
}
